Add documentation for new wire format (#149)
endlessm/azafea: docs/source/events.rst

Setup. In that case, the main user account has been set up as a child user,
with no administrator privileges, and with parental controls enabled. A
second user has been created as the administrator. The event contains the
values of the initial parental controls settings, but no identifying
information about the user (such as their username or full name). The
intention is to allow comparisons of which parental controls are enabled
initially by users, and which are enabled long term.

See `T28741 <https://phabricator.endlessm.com/T28741#810046>`_ and `#101
<https://github.com/endlessm/azafea/pull/101#discussion_r402610338>`_.

The fields in the payload have the same semantics as the properties in the
`AppFilter
<https://gitlab.freedesktop.org/pwithnall/malcontent/-/blob/master/accounts-service/com.endlessm.ParentalControls.AppFilter.xml>`_
interface.

:UUID name: ``MCT_PARENTAL_CONTROLS_EVENT`` in malcontent and gnome-initial-setup

.. versionadded:: 3.8.0

Windows App Opened
~~~~~~~~~~~~~~~~~~

A user tries to open a ``.exe`` or ``.msi`` file.

:UUID name: ``WINDOWS_APP_OPENED`` in eos-gates

.. versionadded:: 2.1.5

Time Spent In Session
~~~~~~~~~~~~~~~~~~~~~

> Time spent in session by all computer users for given .

Reviewer: ```suggestion
Time spent in unlocked user session by a particular local user, within the given aggregation period.
```

.. versionadded:: 4.0.0

Time Foreground App Is Opened
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Duration of an app open during last day.

Reviewer: ```suggestion
Time in seconds (?? minutes ??) that the given app was open in the foreground of an unlocked session within the given aggregation period.
```

Author: I’ll keep seconds, it’s probably useful for some apps that are open for a short amount of time. I don’t think that it’s particularly bad for privacy.

.. versionadded:: 4.0.0

Different Users
~~~~~~~~~~~~~~~

> Number of different users who have logged in during last week.

Reviewer: ```suggestion
Number of different users who have logged in during the given aggregation period.
```

Social Bar Is Visible
~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: It was removed years ago… 3.2.0 I think.

- Payload of start event: ``NULL``
- Payload of stop event: ``NULL``

Desktop Searches
~~~~~~~~~~~~~~~~

Removed in 4.0.0.

We record searches made from the desktop search bar.

- Since: 2.1.2
- UUID: ``b02266bc-b010-44b2-ae0f-8f116ffa50eb``
- UUID name: ``EVENT_DESKTOP_SEARCH`` in gnome-shell
- Payload: type ``(us)``

  - search provider

    - local: 0
    - Google: 1

  - query string (what the user searched for)

.. note::

   Since the 2.1.6. release, Google searches have no longer been recorded.

Knowledge App Search
~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: I'd prefer "not stored since Endless OS 4.0.0". We are unlikely to update the 6 different versions of the Endless SDK (Flatpak) which sends this, but instead just drop it on receipt.

(In fact we never have extracted this event; they're all sitting in the unknown_singular_event table in raw gvariant form.)

We record the search terms used for searching within the knowledge apps along
with the app ID of the knowledge app. (We also record the search term used when
a user performed a desktop search and clicked through to a knowledge app.)

- Since: 2.3.0
- UUID: ``a628c936-5d87-434a-a57a-015a0f223838``
- UUID name: ``SEARCH_METRIC_EVENT_ID`` in eos-knowledge-lib
- Payload: type ``(ss)``

  - search terms
  - application ID

Link Shared from Knowledge App
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: Ditto above

Reported when a user shares a link from one of our content apps on a social network.

- Since: SDK 2
- UUID: ``6775771a-afe7-4158-b7bb-6296fcc7b70d``
- UUID name: ``SHARE_METRIC_EVENT_ID`` in eos-knowledge-lib
- Payload: type ``(sayssu)``

  - Application ID (e.g. ``com.endlessm.animals.en``)
  - ID of content record as a byte array of length 20
    (``ekn://043fd69fe153ac69a05000b60bfea9cff110f14c`` becomes ``[0x04, 0x3f,
    0xd6, 0x9f, 0xe1, 0x53, 0xac, 0x69, 0xa0, 0x50, 0x00, 0xb6, 0x0b, 0xfe,
    0xa9, 0xcf, 0xf1, 0x10, 0xf1, 0x4c]``)
  - Content title
  - The exact URL of the content online, as it was shared to the social network
  - A numerical code indicating which social network the content was shared to

    - Facebook: 0
    - Twitter: 1
    - Whatsapp: 2

See `T18524 <https://phabricator.endlessm.com/T18524>`_.

Knowledge App – Article Open/Close
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: Ditto above

We record when an article is opened or closed in a knowledge app. We record the
ID of the content, the entry point (whether the article was accessed via
article link, desktop search, app link click, or a nav button), the app ID, the
article title, and the content type.

- Since: SDK 2
- UUID: ``fae00ef3-aad7-44ca-aff2-16555e45f0d9``
- UUID name: ``CONTENT_ACCESS_METRIC_EVENT_ID`` in eos-knowledge-lib
- Payload of start event: type ``(ssss)``

  - Entry point
  - Application ID
  - Title
  - Content type

- Payload of stop event: ``NULL``

See `T18516 <https://phabricator.endlessm.com/T18516>`_.

Hack Toolbox - Code View Error - Hack Episode 1
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

- Since: 3.5.3
- UUID: ``e98aa2b8-3f11-4a25-b8e9-b10a635df121``
- UUID name: ``CODEVIEW_ERROR_EVENT``
- Payload of start event: type ``sssa(suquq)``

  - Application ID that this toolbox belongs to (string)
  - ID of function that was being edited in the toolbox (string) (currently always blank)
  - Contents of code view (string)
  - List of the error messages that are displayed, each containing:
  - Text of error message (string)
  - Start line number where error is shown, 1-based (32-bit unsigned)
  - Start column number where error is shown, 0-based (16-bit unsigned)
  - End line number where error is shown, 1-based (32-bit unsigned)
  - End column number where error is shown, 0-based (16-bit unsigned)

- Payload of progress and stop events: type ``s``

  - Diff of the contents of the code view to the state from the previous event,
    in the form of an `ed script
    <https://www.gnu.org/software/diffutils/manual/html_node/ed-Scripts.html>`_,
    chosen because it's the shortest form that the ``diff`` utility can output

See `T24429 <https://phabricator.endlessm.com/T24429>`_.

Hack Clubhouse - Quest
~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

As it's still unclear why that happens, we now detect those situations and
correct them when they happen, so that the metrics system can still be used
afterwards.

:UUID name: ``CACHE_HAS_INVALID_ELEMENTS_EVENT_ID`` in eos-event-recorder-daemon

.. versionadded:: 3.0.9

Discovery Feed Clicked
~~~~~~~~~~~~~~~~~~~~~~

Removed in 4.0.0.

Something is clicked on the Discovery Feed, including content which is not "clickable".

The payload tells us about what users are clicking on generally and whether
they are clicking on things that we don't expect to be clicked. The
intention is to allow the operator to determine what content should be made
clickable and what kinds of content are being opened.

The ``content_type`` field that can be included in the payload is currently
one of:

- ``knowledge_content``: text-based article in a knowledge-app
- ``knowledge_video``: video based article in a knowledge-app
- ``knowledge_artwork``: "artwork" card (larger images in a "Gallery" style format)
- ``undefined``: user clicked on a non-clickable area

:UUID name: ``EVENT_DISCOVERY_FEED_CLICK`` in eos-discovery-feed

.. versionadded:: 3.2.0

Discovery Feed Closed
~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: Not reported since 3.9.0 when the Discovery Feed was removed from Endless OS.

Something is clicked on the Discovery Feed, including content which is not "clickable".

The payload tells us about what users are clicking on generally and whether
they are clicking on things that we don't expect to be clicked. The
intention is to allow the operator to determine what content should be made
clickable and what kinds of content are being opened.

The ``content_type`` field that can be included in the payload is currently
one of:

- ``knowledge_content``: text-based article in a knowledge-app
- ``knowledge_video``: video based article in a knowledge-app
- ``knowledge_artwork``: "artwork" card (larger images in a "Gallery" style format)
- ``undefined``: user clicked on a non-clickable area

:UUID name: ``EVENT_DISCOVERY_FEED_CLICK`` in eos-discovery-feed

.. versionadded:: 3.2.0

Discovery Feed Opened
~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: Not reported since 3.9.0 when the Discovery Feed was removed from Endless OS.

The Discovery Feed is open.

The payload tells us about the language the user is using when interacting
with the discovery feed and how they opened it. The intention is to allow
the operator to determine which languages the discovery feed is most
popular in and how users generally open the feed.

:UUID name: ``EVENT_DISCOVERY_FEED_OPEN`` in eos-discovery-feed

.. versionadded:: 3.2.0

Disk Space Extra
~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Removed in 4.0.0, along with split-disk support in general.
```

Total, used and free disk space for ``/var/endless-extra``.

Sent on startup, and every 24 hours.

See `T18445 <https://phabricator.endlessm.com/T18445>`_.

:UUID name: ``EXTRA_DISK_SPACE_EVENT`` in eos-metrics-instrumentation

.. versionadded:: 3.4.3

Entered Demo Mode
~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since demo mode was removed in Endless OS 3.9.0
```

The systems enters demo mode.

Note that the machine ID will be reset just before the system enters demo
mode, so any metrics collected for this machine ID after this event has
been fired are metrics for a system that is in demo mode.

Demo mode was removed in EOS 3.9, so this metric will not be seen for
systems on 3.9.0 or later.

See `T18983 <https://phabricator.endlessm.com/T18983>`_.

:UUID name: ``DEMO_MODE_ENTERED_METRIC`` in gnome-initial-setup

.. versionadded:: 3.2.5

Endless Application Unmaximized
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: This is another "never going to update the SDK to stop sending this" case – but we can stop storing it.

An in-house application is unmaximized for the first time in each run.

We record a metric with the application ID.

:UUID name: ``UNMAXIMIZE_EVENT`` in eos-sdk

.. versionadded:: 3.0.0

Hack Clubhouse Achievement
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Achievement reached in the Hack Clubhouse.

See `T28004 <https://phabricator.endlessm.com/T28004>`_.

:UUID name: ``ACHIEVEMENT_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse Achievement Points
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Points earned in the Hack Clubhouse.

See `T28004 <https://phabricator.endlessm.com/T28004>`_.

:UUID name: ``ACHIEVEMENT_POINTS_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse Change Page
~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Page changed in the Hack Clubhouse.

See `T28004 <https://phabricator.endlessm.com/T28004>`_.

:UUID name: ``CLUBHOUSE_SET_PAGE_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse Enter Pathway
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Pathway entered in the Hack Clubhouse.

See `T28004 <https://phabricator.endlessm.com/T28004>`_.

:UUID name: ``CLUBHOUSE_PATHWAY_ENTER_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse Mode
~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Hack mode changed in the Hack Clubhouse.

See `T28501 <https://phabricator.endlessm.com/T28501>`_.

:UUID name: ``HACK_MODE_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse News Quest Link
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

Quest link clicked in the Hack Clubhouse news.

See `T29192 <https://phabricator.endlessm.com/T29192>`_.

:UUID name: ``CLUBHOUSE_NEWS_QUEST_LINK_EVENT``

.. versionadded:: 3.7.4

Hack Clubhouse Progress
~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Not reported since Hack was moved to Flathub in Endless OS 3.9.1.
```

    spark. See `T27743 <https://phabricator.endlessm.com/T27743>`_ and
    `T27655 <https://phabricator.endlessm.com/T27655>`_ for more details.

:UUID name: ``USER_LOCATION`` in eos-metrics-instrumentation

.. versionadded:: 2.1.5

Missing Codec
~~~~~~~~~~~~~

Removed in 4.0.0.

A GStreamer-based application tries to install a missing codec.

:UUID name: ``EOS_CODECS_MANAGER_MISSING_CODEC`` in eos-codecs-manager

.. versionadded:: 2.6.0

Monitor Connected
~~~~~~~~~~~~~~~~~

Removed in 4.0.0.

A display is connected (e.g. computer monitor, television) to the machine.

We include any information about the display (e.g. with, height).

:UUID name: ``MONITOR_CONNECTED`` in mutter

UUID was ``566adb36-7701-4067-a971-a398312c2874`` before 2.1.7.

.. versionadded:: 2.1.4

Monitor Disconnected
~~~~~~~~~~~~~~~~~~~~

Removed in 4.0.0.

A display is disconnected (e.g. computer monitor, television) from the machine.

We include any information about the display (e.g. with, height).

:UUID name: ``MONITOR_DISCONNECTED`` in mutter

UUID was ``ce179909-dacb-4b7e-83a5-690480bf21eb`` before 2.1.7.

.. versionadded:: 2.1.4

ShellAppAddedToDesktop
~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Removed in 3.9.0.
```

Shell application is added to desktop.

:UUID name: ``SHELL_APP_ADDED_EVENT`` in gnome-shell

.. note::

    Since 3.9.0 this metric is no longer recorded.
    See `T30661 <https://phabricator.endlessm.com/T30661>`_.

.. versionadded:: 2.1.0

Shell App Removed From Desktop
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Removed in 3.9.0.
```

Shell application is removed from desktop.

:UUID name: ``SHELL_APP_REMOVED_EVENT`` in gnome-shell

.. note::

    Since 3.9.0 this metric is no longer recorded.
    See `T30661 <https://phabricator.endlessm.com/T30661>`_.

.. versionadded:: 2.1.0

Underscan Enabled
~~~~~~~~~~~~~~~~~

> Removed in 4.0.0.

Reviewer: ```suggestion
Removed in 3.9.0.
```
